daemon/src/lib.rs

```diff
@@ -0,0 +1,655 @@
+            match next_restart {
+                Some(time) if now < time => {
+                    event!(Level::ERROR, "{}, waiting until 10 seconds since last restart", status_msg);
+                    sleep(time - now);
+                },
+                _ => (),
+            }
+            next_restart = Some(Instant::now() + Duration::from_secs(10));
+
+            // Restart.
+            event!(Level::INFO, "{}, restarting", status_msg);
+            (child_pid, child_status) = match fork_and_wait_for_startup() {
+                ForkAndWaitResult::ForkFailed { child_pid, status } => (child_pid, Some(status)),
+                ForkAndWaitResult::InParent { child_pid } => (child_pid, None),
+                ForkAndWaitResult::InChild { notify_pipe } => break notify_pipe
+            };
+        }
+    }
+
+    fn should_restart(status: ExitStatus) -> bool {
+        match status.signal() {
+            Some(signal) => {
+                const ERROR_SIGNALS: &[c_int] = &[
+                    libc::SIGABRT, libc::SIGALRM, libc::SIGBUS, libc::SIGFPE, libc::SIGILL,
+                    libc::SIGPIPE, libc::SIGSEGV, libc::SIGXCPU, libc::SIGXFSZ];
+                ERROR_SIGNALS.contains(&signal)
+            },
+            None => false,
+        }
+    }
+
+    fn make_pidfile(pidfile: &Path, cleanup: &mut Cleanup) -> anyhow::Result<()> {
+        // Everyone shares the same file which will be treated as a lock.  To avoid some
+        // uncomfortable race conditions, we can't set up the fatal signal unlink until we've
+        // acquired it.
+        let mut tmpfile = OsString::from(pidfile);
+        tmpfile.push(".tmp");
+        let tmpfile: PathBuf = tmpfile.into();
+
+        let mut file = File::options().append(true).create(true).open(&tmpfile)
+            .with_context(|| format!("{}: create failed", tmpfile.display()))?;
+
+        sys::fcntl_set_lock(&file)
+            .with_context(|| format!("{}: fcntl(F_SETLK) failed", tmpfile.display()))?;
+
+        // We acquired the lock.  Make sure to clean up on exit, and verify
+        // that we're allowed to create the actual pidfile.
+        Self::check_already_running(pidfile)?;
+        cleanup.register_remove_file(pidfile)?;
+
+        file.set_len(0).with_context(|| format!("{}: truncate failed", tmpfile.display()))?;
+
+        file.write_all(format!("{}\n", std::process::id()).as_bytes())
+            .with_context(|| format!("{}: write failed", tmpfile.display()))?;
+
+        std::fs::rename(&tmpfile, &pidfile)
+            .with_context(|| format!("failed to rename {} to {}",
+                                     tmpfile.display(), pidfile.display()))?;
+
+        // Clean up.  Leak 'file' because its file descriptor must remain open to hold the lock.
+        std::mem::forget(file);
+
+        Ok(())
+    }
+
+    fn check_already_running(pidfile: &Path) -> anyhow::Result<()> {
+        match Self::read_pidfile(pidfile, true) {
```

Comment:
this doesn't really check that it's running, just that the pidfile has the specified contents.

Reply:
It's more certain than that, because the pidfile protocol used here will only accept a pidfile as valid if it is `fcntl` locked. If it isn't locked, then `read_pidfile` will treat it as stale. And if a process dies, then the kernel automatically unlocks any `fcntl` locks that it held. It is still possible for there to be problems in special cases (for example, if the pidfile is on a non-POSIX file system such as certain versions of NFS--don't do that please) but those don't come up much.